lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs

```diff
@@ -69,113 +66,245 @@ internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutp
+                        _configurationProvider[evaluatorId]);
+                    Exception e = new NullReferenceException(msg);
+                    Exceptions.Throw(e, Logger);
+                }
+                return GetDataLoadingContextAndServiceConfiguration(partitionDescriptor, evaluatorId);
+            }
+            catch (Exception e)
+            {
+                msg = string.Format("Error while trying to access partition descriptor:{0} from dataset",
+                    _configurationProvider[evaluatorId]);
+                Exceptions.Throw(e, msg, Logger);
+                return null;
+            }
         }
 
-        private ContextAndServiceConfiguration GetUpdateTaskContextAndServiceConfiguration()
+        private ContextAndServiceConfiguration GetDataLoadingContextAndServiceConfiguration(
+            IPartitionDescriptor partitionDescriptor,
+            string evaluatorId)
         {
-            var codecConfig =
+            var dataLoadingContextConf =
+                TangFactory.GetTang()
+                    .NewConfigurationBuilder()
+                    .BindSetEntry<ContextConfigurationOptions.StartHandlers, DataLoadingContext<TDataHandler>, IObserver<IContextStart>>(
+                            GenericType<ContextConfigurationOptions.StartHandlers>.Class,
+                            GenericType<DataLoadingContext<TDataHandler>>.Class)
+                    .Build();
+
+            var serviceConf =
                 TangFactory.GetTang()
-                    .NewConfigurationBuilder(
-                        new[]
-                        {
-                            StreamingCodecConfiguration<MapInputWithControlMessage<TMapInput>>.Conf.Set(
-                                StreamingCodecConfiguration<MapInputWithControlMessage<TMapInput>>.Codec,
-                                GenericType<MapInputWithControlMessageCodec<TMapInput>>.Class).Build(),
-                            StreamingCodecConfigurationMinusMessage<TMapOutput>.Conf.Build(),
-                            _configurationManager.UpdateFunctionCodecsConfiguration
-                        })
+                    .NewConfigurationBuilder(ServiceConfiguration.ConfigurationModule.Build(),
+                        dataLoadingContextConf,
+                        partitionDescriptor.GetPartitionConfiguration())
                     .Build();
 
-            var serviceConf = Configurations.Merge(_groupCommDriver.GetServiceConfiguration(), codecConfig);
-            return new ContextAndServiceConfiguration(_groupCommDriver.GetContextConfiguration(), serviceConf);
+            var contextConf = ContextConfiguration.ConfigurationModule
+                .Set(ContextConfiguration.Identifier, string.Format("DataLoading-{0}", evaluatorId))
+                .Build();
+            return new ContextAndServiceConfiguration(contextConf, serviceConf);
+        }
+
+        private bool IsMaster(string evaluatorId)
```

Comment:
So I can collate to one method probably for checking if it's master. But in general we need to keep tab of master separately from other evaluators right?